### DISCUSSION TITLE
Running btrace agent with external libraries

### DISCUSSION THREAD

**Discussion by kushal-ti** [btraceio/btrace#551] [Q&A]
I'm trying to run Btrace agent with log4j as an external library. I essentially need to use the MDC implementation in log4j to print out some of the contextual data from MDC in the btrace output. 
It's been mentioned here https://github.com/btraceio/btrace/discussions/544#discussioncomment-2870442 and here https://github.com/btraceio/btrace/issues/85#issuecomment-53779651 that btrace in trusted mode can use external libraries as well. However, i found no actual examples or instructions anywhere of using external libraries with Btrace

If I have the following Btrace script

```java
import org.openjdk.btrace.core.annotations.*;
import static org.openjdk.btrace.core.BTraceUtils.*;

import org.openjdk.btrace.core.types.*;
import org.apache.log4j.MDC;

@BTrace(trusted = true)
public class TrustedBTrace {

  @OnMethod(clazz = "/com.*/", method = "/.*/")
  public static void onMethod(@ProbeClassName String className, @ProbeMethodName String methodName, AnyType[] args) {
      String MDCValue = (String) MDC.get("MDCkey");
      System.out.printf("{ \"class\": \"%s\", \"context\": \"%s\" }%n", className, MDCValue);
  }
}
```

This script is successfully compiled by running 
```bash
 btracec -cp '.:log4j-1.2.17.jar' TrustedBTrace.java
```

However, trying to run this script with the btrace-agent with this command 
```bash
 java -noverify -cp ".:log4j-1.2.17.jar" -javaagent:${BTRACE_HOME}/libs/btrace-agent.jar=stdout=true,script=TrustedBTrace.class -jar MyApp.jar
```

fails with the error message 
```bash
java.lang.NoClassDefFoundError: org/apache/log4j/MDC
        at com.codenation.cap.cg_finders_pipeline.Handler.<clinit>(CgFindersPipelineApplication.java)
        at com.codenation.cap.cg_finders_pipeline.CgFindersPipelineApplication.main(CgFindersPipelineApplication.java:38)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        .... 9 more
Caused by: java.lang.ClassNotFoundException: org.apache.log4j.MDC
        at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:445)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:587)
        at org.springframework.boot.loader.LaunchedURLClassLoader.loadClass(LaunchedURLClassLoader.java:151)
        .... ( more
```
Although the stack trace is a little funny, in that it includes classes from my application instead of btrace, at it's core the problem seems to be that the class loader cannot find log4j class, even though it's been included in the class path when running the jar

How do i run the btrace agent so that it can load external jars at run time

**Comment by jbachorik** [comment]
The problem might be caused by the fact that you are matching some spring bootstrap classes and they are loaded by a specific classloader which might not be affected by the `-cp` option. 
If you could restrict the class filter only to your application classes it might help a bit. 

As I mentioned previously I am considering some improvements this area since the current implementation more it less requires any class referenced from the probe handler method to be loaded by the bootstrap classloader as the class filter can match even various JDK methods which themselves are loaded by the bootstrap classloader.

**Reply by kushal-ti** [reply]
> If you could restrict the class filter only to your application classes it might help a bit.

hi, i'm new to Java, so, maybe i'm confused here, would really appreciate if you could clear a few things up. 
as far as i understand, the error occurs because the btrace-agent.jar does not have access to log4j.jar at runtime. I don't know if the java agent being run in the command has access to the same class path/jars as the actuall application.

Is it possible the btrace agent uses its own class loader ? If so can i specify the jars i want the agent to use using some other arguments? Like i said, if you have some kind of reference or instructions on using external libraries along with the btrace agent i would really appreciate that, because as of now, i'm mostly trying to infer how to add the required jars by going through the source code, and i'm not having a lot of success in it

**Reply by jbachorik** [reply]
The BTrace handler code is injected into the instrumented method and is run in its context. 
Therefore, if the instrumented method has access to the `MDC` class then also the handler code does. 

You can add to the bootstrap classpath and system classpath using the agent arguments (https://github.com/btraceio/btrace/wiki#starting-application-with-btrace). You could try the bootstrap path option first as the system classpath might not be considered by the spring classloaders (not sure - I don't know spring internals that well) and you might end up with duplicated class definitions leading to 'funny' situations.

**Comment by kushal-ti** [comment]
This solved my issue thank you

**Reply by jbachorik** [reply]
Glad it helped!
